src/contracts/core/AllocationManager.sol

```diff
@@ -0,0 +1,867 @@
+
+import "../permissions/Pausable.sol";
+import "../libraries/EIP1271SignatureUtils.sol";
+import "../libraries/SlashingLib.sol";
+import "./AllocationManagerStorage.sol";
+
+contract AllocationManager is
+    Initializable,
+    OwnableUpgradeable,
+    Pausable,
+    AllocationManagerStorage,
+    ReentrancyGuardUpgradeable
+{
+    using Snapshots for Snapshots.History;
+
+    /// @dev BIPS factor for slashable bips
+    uint256 internal constant BIPS_FACTOR = 10_000;
+
+    /// @dev Maximum number of pending updates that can be queued for allocations/deallocations
+    /// Note this max applies for a single (operator, strategy, operatorSet) tuple.
+    uint256 public constant MAX_PENDING_UPDATES = 1;
+
+    /// @dev Returns the chain ID from the time the contract was deployed.
+    uint256 internal immutable ORIGINAL_CHAIN_ID;
+
+    /**
+     *
+     *                         INITIALIZING FUNCTIONS
+     *
+     */
+
+    /**
+     * @dev Initializes the immutable addresses of the strategy mananger, delegationManager, slasher,
+     * and eigenpodManager contracts
+     */
+    constructor(
+        IDelegationManager _delegation,
+        IAVSDirectory _avsDirectory,
+        uint32 _DEALLOCATION_DELAY,
+        uint32 _ALLOCATION_DELAY_CONFIGURATION_DELAY
+    )
+        AllocationManagerStorage(_delegation, _avsDirectory, _DEALLOCATION_DELAY, _ALLOCATION_DELAY_CONFIGURATION_DELAY)
+    {
+        _disableInitializers();
+        ORIGINAL_CHAIN_ID = block.chainid;
+        DEALLOCATION_DELAY = _DEALLOCATION_DELAY;
+    }
+
+    /**
+     * @dev Initializes the addresses of the initial owner, pauser registry, and paused status.
+     * minWithdrawalDelayBlocks is set only once here
+     */
+    function initialize(
+        address initialOwner,
+        IPauserRegistry _pauserRegistry,
+        uint256 initialPausedStatus
+    ) external initializer {
+        _initializePauser(_pauserRegistry, initialPausedStatus);
+        _DOMAIN_SEPARATOR = _calculateDomainSeparator();
+        _transferOwnership(initialOwner);
+    }
+
+    /**
+     * @notice Called by the delagation manager to set delay when operators register.
+     * @param operator The operator to set the delay on behalf of.
+     * @param delay The allocation delay in seconds.
+     * @dev msg.sender is assumed to be the delegation manager.
+     */
+    function setAllocationDelay(address operator, uint32 delay) external {
+        require(msg.sender == address(delegation), OnlyDelegationManager());
+        _setAllocationDelay(operator, delay);
+    }
+
+    /**
+     * @notice Called by operators to set their allocation delay.
+     * @param delay the allocation delay in seconds
+     * @dev msg.sender is assumed to be the operator
+     */
+    function setAllocationDelay(
+        uint32 delay
+    ) external {
+        require(delegation.isOperator(msg.sender), OperatorNotRegistered());
+        _setAllocationDelay(msg.sender, delay);
+    }
```

Comment:
should these two (overloaded) functions perhaps be combined? or is that too confusing? I'm imagining something like
```
    function setAllocationDelay(address operator, uint32 delay) external {
       if (msg.sender != address(delegation)) {
          require(msg.sender == operator && delegation.isOperator(msg.sender))
      }
        _setAllocationDelay(operator, delay);
    }
```

Reply:
Will fix.

Reply:
IMO the overloaded functions are fine. It feels a little weird to tell operators they need to pass in their own address otherwise.

Generally it is also strange to have this method's entry point be 2 different contracts, but i'm not sure what satisfactory way there is to solve that.

Reply:
Part of broader open question here of delegated accounts/key rotation. Lets revisit this soon